crypto/cipher_extra/e_chacha20poly1305.c

```diff
@@ -346,3 +382,320 @@ const EVP_AEAD *EVP_aead_chacha20_poly1305(void) {
+
+    // Initialize the poly1305 context
+    CRYPTO_poly1305_init(poly_ctx, poly1305_key);
+    cipher_ctx->key.counter_nonce[0] = 1;
+    cipher_ctx->key.partial_len = 0;
+    cipher_ctx->len.aad = 0;
+    cipher_ctx->len.text = 0;
+    cipher_ctx->poly_initialized = 1;
+  }
+
+  // Handle an |EVP_CipherUpdate|
+  if (inp) {
+    if (out == NULL) {
+      // NULL |out| signals an AAD update
+      CRYPTO_poly1305_update(poly_ctx, inp, inl);
+      cipher_ctx->len.aad += inl;
+      cipher_ctx->pad_aad = 1;
+      return (int32_t) inl;
+    } else {
+      // Finish AAD by applying padding
+      if (cipher_ctx->pad_aad) {
+        remainder = cipher_ctx->len.aad % POLY1305_TAG_LEN;
+        if (remainder != 0) {
+          static const uint8_t padding[POLY1305_TAG_LEN] = {0};
+          CRYPTO_poly1305_update(poly_ctx, padding, sizeof(padding) - remainder);
+        }
+        cipher_ctx->pad_aad = 0;
+      }
+      // cipher/plain text |EVP_CipherUpdate|
+      if (EVP_CIPHER_CTX_encrypting(ctx)) {
+        // Encryption
+        cipher_chacha20_do_cipher(ctx, out, inp, inl);
+        // Update poly1305 with computed ciphertext
+        CRYPTO_poly1305_update(poly_ctx, out, inl);
+        cipher_ctx->len.text += inl;
+      } else {
+        // Decryption
+        // Update poly1305 with incoming ciphertext
+        CRYPTO_poly1305_update(poly_ctx, inp, inl);
+        cipher_chacha20_do_cipher(ctx, out, inp, inl);
+        cipher_ctx->len.text += inl;
+      }
+    }
+  }
+
+  // Process an |EVP_CipherFinal|
+  if (inp == NULL) {
+    uint8_t temp[POLY1305_TAG_LEN];
+    static const uint8_t padding[POLY1305_TAG_LEN] = {0};
+
+    // Finish AAD inp case there were no intermediate Update() calls
+    if (cipher_ctx->pad_aad) {
+      remainder = cipher_ctx->len.aad % POLY1305_TAG_LEN;
+      if (remainder != 0) {
+        CRYPTO_poly1305_update(poly_ctx, padding, sizeof(padding) - remainder);
+      }
+      cipher_ctx->pad_aad = 0;
+    }
+
+    // Apply padding for the text
+    remainder = cipher_ctx->len.text % POLY1305_TAG_LEN;
+    if (remainder != 0) {
+      CRYPTO_poly1305_update(poly_ctx, padding, sizeof(padding) - remainder);
+    }
+
+    // Apply len values
+    CRYPTO_poly1305_update(poly_ctx, (uint8_t *) &cipher_ctx->len,
+                           POLY1305_TAG_LEN);
+
+    // Compute the tag and write it to scratch or the cipher context
+    CRYPTO_poly1305_finish(poly_ctx, EVP_CIPHER_CTX_encrypting(ctx) ?
+      cipher_ctx->tag : temp);
+    cipher_ctx->poly_initialized = 0;
+
+    // Check the tags if we're decrypting
+    if (!EVP_CIPHER_CTX_encrypting(ctx)) {
+      if (CRYPTO_memcmp(temp, cipher_ctx->tag, cipher_ctx->tag_len)) {
+        return -1;
+      }
+    }
+  }
+  return (int32_t) inl;
+}
+
+static void cipher_chacha20_poly1305_cleanup(EVP_CIPHER_CTX *ctx) {
+  CIPHER_CHACHA_POLY_CTX *cipher_ctx = CCP_CTX(ctx);
+  if (cipher_ctx) {
+    OPENSSL_cleanse(ctx->cipher_data, sizeof(CIPHER_CHACHA_POLY_CTX));
+  }
```

Comment:
This looks a little weird, you're checking cipher_ctx but freeing ctx->cipher_data.